frontend/catalyst/compilation_pipelines.py

```diff
@@ -550,9 +546,107 @@ def __call__(self, *args, **kwargs):
         else:
             args = CompiledFunction.promote_arguments(self.c_sig, r_sig, *args)
 
+        if any(isinstance(arg, jax.core.Tracer) for arg in args):
+            # Only compile a derivative version of the compiled function when needed.
+            if self.jaxed_qfunc is None:
+                self.jaxed_qfunc = JAX_QJIT(self)
+
+            return self.jaxed_qfunc(*args, **kwargs)
+
         return self.compiled_function(*args, **kwargs)
 
 
+class JAX_QJIT:
+    """Wrapper class around :class:`~.QJIT` that enables compatibility with JAX transformations.
+
+    The primary mechanism through which this is effected is by wrapping the invocation of the QJIT
+    object inside a JAX ``pure_callback``. Additionally, a custom JVP is defined in order to support
+    JAX-based differentiation, which is itself a ``pure_callback`` around a second QJIT object which
+    invokes :func:`~.grad` on the original function. Using this class thus incurs additional
+    compilation time.
+
+    Args:
+        qfunc (QJIT): the compiled quantum function object to wrap
+    """
+
+    def __init__(self, qfunc):
+        @jax.custom_jvp
+        def jaxed_qfunc(*args, **kwargs):
+            results = self.wrap_callback(qfunc, *args, **kwargs)
+            if len(results) == 1:
+                results = results[0]
+            return results
+
+        self.qfunc = qfunc
+        self.jaxed_qfunc = jaxed_qfunc
+        self._deriv_qfunc = None  # compiled on demand
+        jaxed_qfunc.defjvp(self.compute_jvp)
+
+    @property
+    def deriv_qfunc(self):
+        """Compile a function computing the derivate of the wrapped QJIT function."""
+        if self._deriv_qfunc is not None:
+            return self._deriv_qfunc
+
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(self.qfunc, argnum=all_diff_args)(*args, **kwargs)
```

Comment:
That should still work, although inefficiently. Catalyst will compute the Jacobian for all arguments and sum up JVPs across arguments to produce the `custom_jvp` result. 

Originally, I didn't know if there was a way to determine how the `argnums` (which user may have specified at a higher level) propagated down to the arguments of our function, so this was my first attempt to get it working.
I'm not 100% sure, but if I'm not mistaken arguments that don't participate in differentiation will result in simply having a tangent vector of all zeros for that argument. My assertion here encodes this assumption:
https://github.com/PennyLaneAI/catalyst/blob/c1f649449c63472dc3deb449f7e5a7abcf386c38/frontend/catalyst/compilation_pipelines.py#L628-L629
In any case, a tangent vector of all zeros indicates that the corresponding argument can be skipped for Jacobian computation since the Jacobian-vector product will be zero and thus contribute nothing to the final sum.
 
I actually quickly prototyped this yesterday as follows:
```py
        # Optimization: Do not compute Jacobians for tangents which do not participate in
        #               differentiation.
        # TODO: replace with symbolic zero support
        argnums = []
        for idx, tangent in enumerate(tangents):
            if tangent.dtype.kind == "f":
                if is_abstract(tangent, like="jax"):
                    argnums.append(idx)
                elif any(jnp.atleast_1d(tangent != 0.0)):
                    argnums.append(idx)
        derivatives = self.wrap_callback(self.get_derivative_qfunc(argnums), *primals)
```
and
```py
    def get_derivative_qfunc(self, argnums):
        """Get or generate a derivative version of the wrapped QJIT for a given set of argnums."""

        argnum_key = "".join(str(idx) for idx in argnums)
        if not argnum_key in self.deriv_qfuncs:

            @functools.wraps(self.qfunc, assigned=("__annotations__",), updated=())
            def deriv_wrapper(*args, **kwargs):
                # Only consider differentiable arguments (i.e. of type floating point).
                if any(args[idx].dtype.kind != "f" for idx in argnums):
                    raise ValueError("Attempted to differentiate non-floating point argument!")

                return catalyst.grad(self.qfunc, argnum=argnums)(*args, **kwargs)

            deriv_wrapper.__name__ = "deriv_" + argnum_key + "." + self.qfunc.__name__
            self.deriv_qfuncs[argnum_key] = QJIT(deriv_wrapper, self.qfunc.compile_options)

        return self.deriv_qfuncs[argnum_key]
```
One downside is that we compile more often (every time a new unique set of argnums is encountered), but that should be well worth it for the savings in gradient compute time.

The above can be improved further with something @rmoyard pointed out, which is that JAX recently implemented [symbolic zeros](https://github.com/google/jax/issues/11208) for tangent vectors. Once we upgrade our JAX version we could make use of it here.

Reply:
I implemented the optimization here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/f89faf6b71d41069e2f44593b4edf9fd03a36916